test/autostop.coffee

```diff
@@ -0,0 +1,497 @@
+            @fixture2 = getFixture()
+            @fixture3 = getFixture()
+            @fixture4 = getFixture()
+            @fixture2.slug = 'test2'
+            @fixture3.slug = 'test3'
+            @fixture3.isStoppable = false
+            @fixture4.slug = 'test4'
+            @fixture4.state = 'stopped'
+
+            @sandbox
+                .stub Application, 'all'
+                .callsArgWithAsync 0, null, [@fixture1, @fixture2, @fixture3, @fixture4]
+
+            @startTimeout = @sandbox.stub AutostopManager, 'startTimeout'
+
+        after -> @sandbox.restore()
+
+        it "When .init() is called", (done) ->
+            AutostopManager.init done
+
+        it "Then all auto-stoppable running applications should have a timeout running", ->
+            @startTimeout.callCount.should.equal 2
+
+
+    describe "Functional test", ->
+
+        before helpers.setup TESTPORT
+        before helpers.createUser 'test@cozycloud.cc', 'testtest'
+
+        #initiate fake servers
+        before ->
+            @controller = helpers.fakeServer { drone: { port: 8001 } }, 200
+            @controller.listen 9002
+            @proxy = helpers.fakeServer msg: "ok", 200
+            @proxy.listen 9104
+
+        before ->
+            @client = helpers.getClient TESTPORT, @
+            @sandbox = sinon.sandbox.create useFakeTimers: true
+
+        after -> @sandbox.restore()
+        after ->
+            @controller.close()
+            @proxy.close()
+        after helpers.takeDown
+
+        it "When I start an auto-stoppable application", (done) ->
+            creator = helpers.createApp "Test App", "test-app", 0, "installed"
+            creator done
+
+        it "Then I wait for 4 minutes", ->
+            @sandbox.clock.tick 4 * minute
+
+        it "Then it should still be running", (done) ->
+            Application.all key: 'test-app', (err, apps) ->
+                should.not.exist err
+                should.exist apps
+                apps[0].state.should.equal 'installed'
+                done()
+
+        it "Then the application does a request to the Data System", (done) ->
+            ds = new Client "http://localhost:9101/"
+            ds.setBasicAuth 'test-app', 'test-app'
+            data =
+                docType: 'Event'
+                description: 'blablabla'
+            ds.post 'data/', data, (err, res, body) ->
+                done()
+
+        it "Then I wait for 4 minutes", ->
+            @sandbox.clock.tick 4 * minute
+
+        it "Then it should still be running", ->
+            Application.all key: 'test-app', (err, apps) ->
+                should.not.exist err
+                should.exist apps
+                apps[0].state.should.equal 'installed'
+
+        it "Then I wait for one more minute", ->
+            @sandbox.clock.tick 2 * minute
+            @sandbox.clock.restore()
+
+        it "Then the application should be stopped", (done) ->
+            @timeout 6000
```

Comment:
I do the change. It's okay to use a timeout because I restore the setTimeout mocking in the previous step.